### PR TITLE
End span after closing scope

### LINKED ISSUE #4596: Update instrumentation to always close scope before ending span
From https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4544#discussion_r743230890

ByteBuddy instrumentation (`@OnMethodEnter/@OnMethodExit`) follows this pattern, so would be good for consistency for other instrumentation to follow it also. In particular, `Context.current()` would be consistent in all end hooks, e.g. SpanProcessor or Instrumenter.

```
  Response response;
  try (Scope scope = context.makeCurrent()) {
    response = actualMethod(request);
  } catch (Throwable t) {
    instrumenter.end(context, request, null, t);
  }
  instrumenter.end(context, request, response, null);
  return response;
```

If we make this change we should update the Instrumenter documentation linked above also.

### PR DESCRIPTION
**Pull request by trask**
Resolves #4596

### CODE REVIEW

File: instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/internal/TracingInterceptor.java [hunk 1]
```diff
@@ -41,17 +41,15 @@ public Response intercept(Chain chain) throws IOException {
...
+    Response response;
     try (Scope ignored = context.makeCurrent()) {
       response = chain.proceed(request);
-      return response;
-    } catch (Exception e) {
-      error = e;
-      throw e;
-    } finally {
-      instrumenter.end(context, chain, response, error);
+    } catch (Exception t) {
```
Comment (laurit) [inline instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/internal/TracingInterceptor.java:47]:
```suggestion
    } catch (Throwable t) {
```

File: instrumentation/spring/spring-webmvc/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcTelemetryProducingFilter.java [hunk 2]
```diff
@@ -60,21 +60,20 @@ public void doFilterInternal(
     }
 
     Context context = instrumenter.start(parentContext, request);
+    if (httpRouteSupport.hasMappings()) {
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcTelemetryProducingFilter.java:63]: I suspect that it was deliberately added after `filterChain.doFilter`.

File: instrumentation/spring/spring-webmvc/spring-webmvc-6.0/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v6_0/WebMvcTelemetryProducingFilter.java [hunk 3]
```diff
@@ -60,21 +60,20 @@ public void doFilterInternal(
     }
 
     Context context = instrumenter.start(parentContext, request);
+    if (httpRouteSupport.hasMappings()) {
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc/spring-webmvc-6.0/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v6_0/WebMvcTelemetryProducingFilter.java:63]: I suspect that it was deliberately added after `filterChain.doFilter`.

File: instrumentation/spring/spring-webmvc/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcTelemetryProducingFilter.java (hunk 2 above)
Comment (trask) [reply instrumentation/spring/spring-webmvc/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcTelemetryProducingFilter.java]: thanks, reverted changes to these two files

**Review by laurit — APPROVED** [review]

**Review by jaydeluca — APPROVED** [review]

**Pull request merged**